sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -2593,52 +2631,406 @@ gather_impl(AccessorTy acc, simd<uint32_t, N> offsets, uint32_t glob_offset,
-/// as a single \ref simd object. An element can be a 1, 2 or 4-byte value.
+/// Collects elements from memory referenced by the accessor \p acc, byte
+/// offsets \p byte_offsets and common offset \glob_offset, then returns
+/// the loaded elements as a single \ref simd object.
 ///
-/// @tparam T Element type; can only be a 1,2,4-byte integer, \c sycl::half or
-///   \c float.
-/// @tparam N The number of vector elements. Can be \c 1, \c 8, \c 16 or \c 32.
-/// @tparam AccessorTy The accessor type.
+/// Supported platforms: DG2/PVC if sizeof(T) > 4 or the number of elements to
+/// load is not equal to 1, 8, 16, 32. Otherwise, it is supported on ALL
+/// platforms.
+///
+/// @tparam T Element type.
+/// @tparam N The number of vector elements.
+/// @tparam AccessorT The accessor type.
 /// @param acc The accessor to gather from.
-/// @param offsets Per-element offsets in bytes.
+/// @param byte_offsets Per-element offsets in bytes.
 /// @param glob_offset Offset in bytes added to each individual element's offset
 ///   to compute actual memory access offset for that element.
 /// @param mask Memory access mask. Elements with zero corresponding mask's
 ///   predicate are not accessed, their values in the resulting vector are
 ///   undefined.
 ///
-template <typename T, int N, typename AccessorTy>
-__ESIMD_API std::enable_if_t<
-    (sizeof(T) <= 4) && (N == 1 || N == 8 || N == 16 || N == 32) &&
-        detail::is_device_accessor_with_v<AccessorTy,
-                                          detail::accessor_mode_cap::can_read>,
-    simd<T, N>>
-gather(AccessorTy acc, simd<detail::DeviceAccessorOffsetT, N> offsets,
-       detail::DeviceAccessorOffsetT glob_offset = 0, simd_mask<N> mask = 1) {
+// Dev note: the argument \p glob_offset of this function does not have
+// the default value to not conflict with more generic variant (acc-ga-3)
+// defined lower. This restriction though requires adding an additional
+// variant: simd<T, N> gather(acc, glob_offset) to support calls that require
+// implicit conversion of a scalar offset to a vector of offsets, e.g.
+// 'res = gather<T, N>(acc, 0);'
+template <typename T, int N, typename AccessorT>
+__ESIMD_API
+    std::enable_if_t<detail::is_device_accessor_with_v<
+                         AccessorT, detail::accessor_mode_cap::can_read>,
+                     simd<T, N>>
+    gather(AccessorT acc, simd<detail::DeviceAccessorOffsetT, N> byte_offsets,
+           detail::DeviceAccessorOffsetT glob_offset, simd_mask<N> mask = 1) {
 #ifdef __ESIMD_FORCE_STATELESS_MEM
   return gather<T, N>(__ESIMD_DNS::accessorToPointer<T>(acc, glob_offset),
-                      offsets, mask);
+                      byte_offsets, mask);
 #else
-  return detail::gather_impl<T, N, AccessorTy>(acc, offsets, glob_offset, mask);
-#endif
+  if constexpr (sizeof(T) > 4 || !((N == 1 || N == 8 || N == 16 || N == 32))) {
+    // Requires DG2 or PVC implementation.
```

Comment:
```suggestion
    // Requires DG2 or PVC.
```

Unclear to me what implementation means here

Reply:
Thank you, I applied your comment.